doc/api/errors.markdown

```diff
@@ -0,0 +1,448 @@
+type. For example, passing a function to a parameter which expects a string would
+be considered a TypeError.
+
+```javascript
+require('url').parse(function() { }); // throws TypeError, since it expected a string
+```
+
+Node will generate and throw TypeError instances *immediately* -- they are a form
+of argument validation.
+
+### Class: ReferenceError
+
+A subclass of Error that indicates that an attempt is being made to access a variable
+that is not defined. Most commonly it indicates a typo, or an otherwise broken program.
+Only V8 may generate and propagate these errors.
+
+```javascript
+doesNotExist; // throws ReferenceError, `doesNotExist` is not a variable in this program.
+```
+
+ReferenceError instances will have an `.arguments` member that is an array containing
+one element -- a string representing the variable that was not defined.
+
+```javascript
+try {
+  doesNotExist;
+} catch(err) {
+  err.arguments[0] === 'doesNotExist';
+}
+```
+
+Unless the userland program is dynamically generating and running code,
+ReferenceErrors should always be considered a bug in the program, or its
+dependencies.
+
+### Class: SyntaxError 
+
+A subclass of Error that indicates that a program is not valid JavaScript.
+These errors may only be generated and propagated as a result of code
+evaluation. Code evaluation may happen as a result of `eval`, `Function`,
+`require`, or [vm](vm.html). These errors are almost always indicative of a broken
+program.
+
+```javascript
+try {
+  require("vm").runInThisContext("binary ! isNotOk");
+} catch(err) {
+  // err will be a SyntaxError
+}
+```
+
+SyntaxErrors are unrecoverable from the context that created them – they may only be caught
+by other contexts.
+
+### Exceptions vs. Errors
+
+<!--type=misc-->
+
+A JavaScript "exception" is a value that is thrown as a result of an invalid operation or
+as the target of a `throw` statement. While it is not required that these values inherit from
+`Error`, all exceptions thrown by Node or the JavaScript runtime *will* be instances of Error.
+
+Some exceptions are *unrecoverable* at the JavaScript layer. These exceptions will always bring 
+down the process. These are usually failed `assert()` checks or `abort()` calls in the C++ layer.
+
+## System Errors
+
+System errors are generated in response to a program's runtime environment.
+Ideally, they represent operational errors that the program needs to be able to
+react to. They are generated at the syscall level: an exhaustive list of error
+codes and their meanings is available by running `man 2 intro` on most Unices;
+or [online](http://man7.org/linux/man-pages/man2/intro.2.html).
+
+In Node, system errors are represented as augmented Error objects -- not full
+subclasses, but instead an error instance with added members.
+
+### Class: System Error
+
+#### error.syscall
+
+A string representing the [syscall](http://man7.org/linux/man-pages/man2/syscall.2.html) that failed.
+
+#### error.errno
```

Comment:
`errno` is just an alias to `code`, per [util._errnoException](https://github.com/iojs/io.js/blob/v1.x/lib/util.js#L735-L736).